ReportUpdater.php

```diff
@@ -22,97 +67,136 @@ public function __construct() {
+			return;
 		}
-		// Update index page
-		$this->api->updateIndex( 'User:Community Tech bot/Popular pages' );
+
+		if ( count( $pages ) > 1000000 ) { // See T164178
+			wfLogToFile( 'Error: ' . $project . ' is too large. Skipping.' );
+			return;
+		}
+
+		$pageviews = $this->api->getMonthlyPageviews(
+			array_keys( $pages ),
+			date( 'Ymd00', $this->start ),
+			date( 'Ymd00', $this->end )
+		);
+
+		// Compute total views for the month
+		$totalViews = array_sum( array_values( $pageviews ) );
+
+		// Truncate to configured limit.
+		$pageviews = array_slice( $pageviews, 0, $config['Limit'], true );
+
+		// Populate new reported pages.
+		foreach ( $pageviews as $title => $views ) {
+			$pageviews[$title] = array_merge( $pages[$title], [
+				'views' => $views,
+				'avgViews' => floor(
+					$views / ( floor( ( $this->end - $this->start ) / ( 60 * 60 * 24 ) ) )
+				),
+			] );
+
+			// Attempt to reduce memory usage, probably unneeded.
+			unset( $pages[$title] );
+		}
+
+		$hasLeadSection = $this->api->hasLeadSection( $config['Report'] );
+
+		// Generate and return wikitext.
+		$output = $this->twig->render( 'report.wikitext.twig', [
+			'hasLeadSection' => $hasLeadSection,
+			'wiki' => $this->wiki,
+			'start' => $this->start,
+			'end' => $this->end,
+			'project' => $project,
+			'pages' => $pageviews,
+			'totalViews' => $totalViews,
+			'assessments' => $this->api->getAssessmentConfig(),
+			'category' => $this->api->getWikiConfig()['category'],
+		] );
+
+		$this->api->setText( $config['Report'], $output, (int)$hasLeadSection );
+	}
+
+	/**
+	 * Update index page for the bot, listing each WikiProject, its report
+	 * and when it was last updated.
+	 */
+	public function updateIndex() {
+		$projectsConfig = $this->api->getJSONConfig();
+
+		foreach ( $projectsConfig as $project => &$data ) {
+			$data['Updated'] = $this->api->getBotLastEditDate( $project );
+		}
+
+		// Generate and return wikitext.
+		$output = $this->twig->render( 'index.wikitext.twig', [
+			'projects' => $projectsConfig,
+			'configPage' => $this->api->getWikiConfig()['config'],
+		] );
+
+		$this->api->setText( $this->api->getWikiConfig()['index'], $output );
+	}
+
+	/**
+	 * Validate the WikiProject config entry, including the syntax within the JSON config,
+	 * whether target page is the correct namespace, and if the target page actually exists.
+	 *
+	 * @param string $project The WikiProject.
+	 * @param array $config The WikiProject's configuration.
+	 * @return bool true if valid, false otherwise.
+	 */
+	private function validateProjectConfig( $project, $config ) {
+		// Check that config values are set
+		if ( !isset( $config['Name'] ) || !isset( $config['Limit'] ) || !isset( $config['Report'] ) ) {
+			wfLogToFile( 'Error: Incomplete data in config for ' . $project . '. Skipping.' );
+			return false;
+		}
+
+		// Make sure report will be written to Wikipedia namespace
+		if ( strpos( $config['Report'], 'Wikipedia:' ) !== 0 ) {
```

Comment:
We probably need to change this as the "Wikipedia:" prefix is English specific.

Reply:
Maybe you could just check that it isn't being written to the main (0) namespace instead. The idea behind this check is to prevent the bot from being used to erase/vandalize articles.

Reply:
@MusikAnimal - Although there's no easy foolproof way to do this, we can reject obviously bad submissions like:
```
// Don't allow writing report to main namespace. There is no easy way to grab the
// namespace ID so just reject titles that don't have a colon in them.
if ( strpos( $config['Report'], ':' ) === false ) {
```
Personally, I think having some kind of check is important (even if it's hacky), mainly to prevent accidentally writing to Main namespace under the assumption that the namespace is automatically set to something.

Reply:
Done! With a future PR we can make this more exacting and use the API to get the namespace names.